plugins/prom.go

```diff
@@ -0,0 +1,101 @@
+package plugins
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"golang.org/x/net/context"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+type PromServer struct {
+	port        int
+	srvr        *http.Server
+	tcpListener *net.TCPListener
+	lggr        logger.Logger
+
+	handler http.Handler
+}
+
+type PromServerOpt func(*PromServer)
+
+func WithRegistry(r *prometheus.Registry) PromServerOpt {
+	return func(s *PromServer) {
+		s.handler = promhttp.HandlerFor(r, promhttp.HandlerOpts{})
+	}
+}
+
+func NewPromServer(port int, lggr logger.Logger, opts ...PromServerOpt) *PromServer {
+
+	s := &PromServer{
+		port: port,
+		lggr: lggr,
+		srvr: &http.Server{},
+
+		handler: promhttp.Handler(),
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	return s
+}
+
+// Start start HTTP server on specified port to handle metrics requests
+func (p *PromServer) Start() error {
+	err := p.setupListener()
+	if err != nil {
+		return err
+	}
+
+	http.Handle("/metrics", p.handler)
+
+	go func() {
+		err := p.srvr.Serve(p.tcpListener)
+		if errors.Is(err, net.ErrClosed) {
+			// ErrClose is expected on gracefully shutdown
+			p.lggr.Warnf("%s closed", p.Name())
+		} else {
+			p.lggr.Errorf("%s: %s", p.Name(), err)
+		}
+	}()
+	return nil
+}
+
+// Close shutdowns down the underlying HTTP server. See [http.Server.Close] for details
+func (p *PromServer) Close() error {
+	return p.srvr.Shutdown(context.Background())
+}
+
+// Name of the server
+func (p *PromServer) Name() string {
+	return fmt.Sprintf("%s-prom-server", p.lggr.Name())
+}
+
+func (p *PromServer) Port() int {
+	// always safe to cast because we explicitly have a tcp listener
+	// doesn't seem to be direct access to Port without the addr casting
+	return p.tcpListener.Addr().(*net.TCPAddr).Port
+
+}
+
+// setupListener creates explicit listener so that we can resolve `:0` port, which is needed for testing
+// if we didn't need the resolved addr, or could pick a static port we could use p.srvr.ListenAndServer
```

Comment:
Would `ResolveTCPAddr` help simplify things? That's what we use here:
https://github.com/smartcontractkit/chainlink/blob/develop/core/services/chainlink/config_general.go#L833-L843

Reply:
Or I just stumbled on this: https://pkg.go.dev/gvisor.dev/gvisor/pkg/tcpip/ports
> Package ports provides PortManager that manages allocating, reserving and releasing ports.

Reply:
I tried ResolveTCPAddr and it doesn't do what we need -- the 0 value port is not assigned until ListenTCP returns.

updated the code comment to reflect this. 

don't think we need an external package here yet.

Reply:
But are we building the same exact thing? I thought it looked like it might drop right in and just work so we don't have to think about or test any of this.

Reply:
the main difficultly is not picking a port, it's picking the same port for the same plugin if that plugin restarts. otherwise the `discovery` will bloat with wrong endpoints and we would have to handle cleanup there.

the external pkg could pick the port for us, but we don't need a package to do that.

currently the ports are deterministic and really just a start and offset. this is sufficient because we control all the processes in the container.

if there is a reason for random ports then the pkg would be useful because it has the notion of a reservation.

Reply:
i read re-read your comment. maybe one thing that is missing between us is the node picks the port and tells the plugin to use it. this addresses the plugin restart issue i mentioned above.

you are right that if the plugin picked the port then we could simplify some of this particular code, and we would definitely need a reservation system to avoid races, but we would also to a way for the plugin to tell the node which port it picked so the node can publish the right discovery url